File: Structural/Bridge/Kotlin/BridgePattern.kt
abstract class Implementor
{
    abstract fun Operation()
}

class ConcreteImplementor1: Implementor()
{
    public override fun Operation()
    {
        println("ConcreteImplementor1::Operation")
    }
}

class ConcreteImplementor2: Implementor()
{
    public override fun Operation()
    {
        println("ConcreteImplementor2::Operation")
    }
}

open class Abstraction
{
    protected var implementor:Implementor? = null

    fun set_Implementor(value:Implementor)
    {
        implementor = value
    }


    open fun Operation() {
        implementor?.Operation()
    }
}

class RefinedAbstraction: Abstraction()
{

    public override fun Operation()
    {
        implementor?.Operation()
    }
}


// test
object Program
{
    @JvmStatic fun main(args:Array<String>)
    {
        val obj = RefinedAbstraction()
        obj.set_Implementor(ConcreteImplementor1())
        obj.Operation()

        val obj2 = RefinedAbstraction()
        obj2.set_Implementor(ConcreteImplementor2())
        obj2.Operation()

    }
}

File: Structural/Bridge/Swift/bridgePattern.swift
protocol Implementor
{
    func Operation()
}

public class ConcreteImplementor1: Implementor 
{
    func Operation()
    {
        print("ConcreteImplementor1::Operation")
    }
}

public class ConcreteImplementor2: Implementor 
{
    func Operation()
    {
        print("ConcreteImplementor2::Operation")
    }
}

protocol Abstraction 
{
    var implementor:Implementor
    {
        get 
        set
    }

    init(implementor:Implementor)
    func Operation()
}

final class RefinedAbstraction: Abstraction
{
    internal var mImplementor: Implementor
    var implementor:Implementor
    {
        get
        {
            return mImplementor
        }
        set(val)
        {
            mImplementor = val
        }
    }

    init(implementor:Implementor)
    {
        self.mImplementor = implementor
    }

    func Operation()
    {
        implementor.Operation()
    }
}

// test
var obj:Abstraction = RefinedAbstraction(implementor: ConcreteImplementor1())
obj.Operation()

var obj2:Abstraction = RefinedAbstraction(implementor: ConcreteImplementor2())
obj2.Operation()
